File: 14-TypeScript/Grey Multi-Tenant/packages/grey-scala/src/main/scala/com/grey/sdk/domain/QueryClient.scala
package com.grey.sdk.domain

import com.grey.sdk.error.GreyError
import com.grey.sdk.grpc.{GreyChannel, QueryData, QueryGrpcService}
import scala.concurrent.{ExecutionContext, Future}

/** Client for generic query operations.
  *
  * Provides a flexible query interface for custom endpoints.
  */
final class QueryClient(
    channel: GreyChannel,
    service: QueryGrpcService
)(using ec: ExecutionContext):

  /** Executes a query against the specified endpoint.
    *
    * @param endpoint
    *   The query endpoint path
    * @param params
    *   Optional query parameters
    * @param requireAuth
    *   Whether authentication is required (default: true)
    * @return
    *   Future containing QueryData on success
    * @throws GreyError
    *   on failure
    */
  def query(
      endpoint: String,
      params: Map[String, String] = Map.empty,
      requireAuth: Boolean = true
  ): Future[QueryData] =
    // Validate inputs
    if endpoint.isEmpty then
      return Future.failed(GreyError.validation("Endpoint is required"))

    // Check authentication if required
    if requireAuth && !channel.isAuthenticated then
      return Future.failed(GreyError.unauthorized())

    service.query(endpoint, params, requireAuth)
      .recoverWith { case e => Future.failed(GreyError.fromThrowable(e)) }

  /** Executes query and returns Either instead of throwing. */
  def querySafe(
      endpoint: String,
      params: Map[String, String] = Map.empty,
      requireAuth: Boolean = true
  ): Future[Either[GreyError, QueryData]] =
    query(endpoint, params, requireAuth)
      .map(Right(_))
      .recover { case e: GreyError => Left(e) }

object QueryClient:
  def apply(channel: GreyChannel, service: QueryGrpcService)(using ExecutionContext): QueryClient =
    new QueryClient(channel, service)

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-dart/lib/src/domain/query_client.dart
/// Query domain client for Grey SDK.
library;

import 'dart:async';

import 'package:grpc/grpc.dart';

import '../error/grey_error.dart';
import '../grpc/grey_channel.dart';
import '../grpc/query_grpc_service.dart';
import '../grpc/types.dart';

/// Client for generic query operations.
class QueryClient {
  /// Creates a new query client.
  QueryClient(this._channel, this._service);

  final GreyChannel _channel;
  final QueryGrpcService _service;

  /// Executes a query against the specified endpoint.
  ///
  /// [endpoint] is the query endpoint path.
  /// [params] are optional query parameters.
  /// [requireAuth] specifies whether authentication is required (default: true).
  ///
  /// Throws [GreyError] on failure.
  Future<QueryData> query({
    required String endpoint,
    Map<String, String>? params,
    bool requireAuth = true,
  }) async {
    // Validate inputs
    if (endpoint.isEmpty) {
      throw GreyError(
        code: GreyError.validationError,
        message: 'Endpoint is required',
      );
    }

    // Check authentication if required
    if (requireAuth && !_channel.isAuthenticated) {
      throw GreyError(
        code: GreyError.unauthorized,
        message: 'Not authenticated',
      );
    }

    try {
      return await _service.query(
        endpoint: endpoint,
        params: params,
        requireAuth: requireAuth,
      );
    } on GrpcError catch (e) {
      throw GreyError.fromGrpcError(e);
    } catch (e) {
      throw GreyError.fromException(e);
    }
  }
}
